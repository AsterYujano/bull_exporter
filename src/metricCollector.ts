import bull from 'bull';
import { Gauge, register as globalRegister, Registry } from 'prom-client';

import { getStats, makeGuages, QueueGauges } from './queueGauges';

export interface BullOptions extends  Pick<bull.QueueOptions, Exclude<keyof bull.QueueOptions, 'redis'>> {
  redis?: string |  bull.QueueOptions['redis'];
}

export class MetricCollector {

  private readonly queues: {
    queue: bull.Queue;
    name: string;
    prefix: string;
  }[];

  private readonly guages: QueueGauges;

  constructor(
    statPrefix: string,
    queueNames: string[],
    opts: BullOptions,
    registers: Registry[] = [globalRegister],
  ) {
    this.queues = queueNames.map(name => ({
      name,
      queue: new bull(name, opts as bull.QueueOptions),
      prefix: opts.prefix || 'bull',
    }));

    this.guages = makeGuages(statPrefix, registers);
  }

  public async updateAll(): Promise<void> {
    const updatePromises = this.queues.map(q => getStats(q.prefix, q.name, q.queue, this.guages));
    await Promise.all(updatePromises);
  }

  public async close(): Promise<void> {
    await Promise.all(this.queues.map(q => q.queue.close()));
  }

}
