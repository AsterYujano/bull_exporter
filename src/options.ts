import yargs from 'yargs';

import { version } from '../package.json';

export interface Options {
  url: string;
  prefix: string;
  metricPrefix: string;
  once: boolean;
  port: number;
  bindAddress: string;
  _: string[];
}

export function getOptionsFromArgs(...args: string[]): Options {
  return yargs
    .usage('prom-metrics [queueNames]')
    .alias('V', 'version')

    .alias('u', 'url')
    .demandOption('url', 'A redis connection url')
    .default('url', 'redis://127.0.0.1:6379')

    .alias('p', 'prefix')
    .demandOption('prefix', 'bull prefix')
    .default('prefix', 'bull')

    .alias('m', 'metric-prefix')
    .demandOption('metric-prefix', 'Metric prefix')
    .default('metric-prefix', 'uhapp_queue_')

    .boolean('once')
    .alias('n', 'once')
    .demandOption('once', 'Print stats and exit without starting a server')
    .default('once', false)

    .number('port')
    .default('port', 5959)

    .alias('b', 'bindAddress')
    .demandOption('bindAddress', 'Address to listen on')
    .default('bindAddress', '0.0.0.0')

    .demandCommand(1)
    .version(version)
    .parse(args) as any;
}
