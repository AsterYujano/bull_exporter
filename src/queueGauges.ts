import bull from 'bull';
import { Gauge, Registry } from 'prom-client';

export interface QueueGauges {
  completed: Gauge;
  active: Gauge;
  delayed: Gauge;
  failed: Gauge;
  waiting: Gauge;
}

export function makeGuages(statPrefix: string, registers: Registry[]): QueueGauges {
  return {
    completed: new Gauge({
      registers,
      name: `${statPrefix}completed`,
      help: 'Number of completed messages',
      labelNames: ['queue', 'prefix'],
    }),
    active: new Gauge({
      registers,
      name: `${statPrefix}active`,
      help: 'Number of active messages',
      labelNames: ['queue', 'prefix'],
    }),
    delayed: new Gauge({
      registers,
      name: `${statPrefix}delayed`,
      help: 'Number of delayed messages',
      labelNames: ['queue', 'prefix'],
    }),
    failed: new Gauge({
      registers,
      name: `${statPrefix}failed`,
      help: 'Number of failed messages',
      labelNames: ['queue', 'prefix'],
    }),
    waiting: new Gauge({
      registers,
      name: `${statPrefix}waiting`,
      help: 'Number of waiting messages',
      labelNames: ['queue', 'prefix'],
    }),
  };
}

export async function getStats(prefix: string, name: string, queue: bull.Queue, gauges: QueueGauges): Promise<void> {
  const { completed, active, delayed, failed, waiting } = await queue.getJobCounts();

  gauges.completed.set({ prefix, queue: name }, completed);
  gauges.active.set({ prefix, queue: name }, active);
  gauges.delayed.set({ prefix, queue: name }, delayed);
  gauges.failed.set({ prefix, queue: name }, failed);
  gauges.waiting.set({ prefix, queue: name }, waiting);
}
