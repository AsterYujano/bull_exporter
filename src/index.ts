import promClient from 'prom-client';
import yargs from 'yargs';

import { version } from '../package.json';

import { MetricCollector } from './metricCollector';

export function getArgs(...args: string[]): { url: string, prefix: string, _: string[] } {
  return yargs
    .usage('prom-metrics [queueNames]')

    .alias('u', 'url')
    .demandOption('url', 'A redis connection url')
    .default('url', 'redis://127.0.0.1:6379')

    .alias('p', 'prefix')
    .demandOption('prefix', 'metrics prefix')
    .default('prefix', 'uhapp_queue_')

    .demandCommand(1)
    .version(version)
    .parse(args) as any;
}

export async function main(...args: string[]): Promise<void> {
  const opts = getArgs(...args);

  const collector = new MetricCollector(opts.prefix, opts._, { redis: opts.url });
  await collector.updateAll();
  await collector.close();
  console.log(promClient.register.metrics());
}

if (require.main === module) {
  const args = process.argv.slice(2);

  let exitCode = 0;
  main(...args)
    .catch(() => process.exitCode = exitCode = 1)
    .then(() => {
      setTimeout(
        () => {
          console.error('No clean exit after 5 seconds, force exit');
          process.exit(exitCode);
        },
        5000,
      ).unref();
    });
}
