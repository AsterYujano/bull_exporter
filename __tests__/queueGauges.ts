import { getStats } from '../src/queueGauges';

import { TestData } from './create.util';
import { getCurrentTestHash } from './setup.util';

let testData: TestData;

beforeEach(async () => {
  jest.resetModuleRegistry();
  const { makeQueue } = await import('./create.util');
  const hash = getCurrentTestHash();
  testData = makeQueue(hash);
});

afterEach(async () => {
  await testData.queue.empty();
  await testData.queue.close();
});

it('should list 1 queued job', async () => {

  const {
    name,
    queue,
    prefix,
    guages,
    registry,
  } = testData;

  await queue.add({ a: 1 });

  await getStats(prefix, name, queue, guages);

  expect(registry.metrics()).toMatchSnapshot();
});
