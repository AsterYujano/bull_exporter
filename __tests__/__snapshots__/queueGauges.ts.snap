// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should list 1 queued job 1`] = `
"# HELP test_stat_completed Number of completed messages
# TYPE test_stat_completed gauge
test_stat_completed{prefix=\\"test-queue\\",queue=\\"54cceb2dc1d55935\\"} 0

# HELP test_stat_active Number of active messages
# TYPE test_stat_active gauge
test_stat_active{prefix=\\"test-queue\\",queue=\\"54cceb2dc1d55935\\"} 0

# HELP test_stat_delayed Number of delayed messages
# TYPE test_stat_delayed gauge
test_stat_delayed{prefix=\\"test-queue\\",queue=\\"54cceb2dc1d55935\\"} 0

# HELP test_stat_failed Number of failed messages
# TYPE test_stat_failed gauge
test_stat_failed{prefix=\\"test-queue\\",queue=\\"54cceb2dc1d55935\\"} 0

# HELP test_stat_waiting Number of waiting messages
# TYPE test_stat_waiting gauge
test_stat_waiting{prefix=\\"test-queue\\",queue=\\"54cceb2dc1d55935\\"} 1
"
`;
